Sm/gauntlet-docs (#499)
forcedotcom/source-deploy-retrieve: contributing/metadata.md

# Contributing Metadata Types to the Registry

> NOTE: If you have a “Settings” metadata type, you do not need to follow these instructions. Support for your type is already built into SDR.

## Adding new types to the registry via a script

To simplify modifying the registry, there's a script.

> > Note: The script depends on being able to create an org that will have your type. If you have a working org but your type isn't compatible with scratch org, you can [use an existing org](#use-an-existing-org)

Reviewer: ```suggestion
> Note: The script depends on being able to create an org that will have your type. If you have a working org, but your type isn't compatible with scratch org, you can [use an existing org](#use-an-existing-org)
```

> 1. looks for missing types (similar to the completeness test)

Reviewer: ```suggestion
The script will:
1. Look for missing types (similar to the completeness test)
```

2. For missing types, generate a project and scratch org that includes the Features/Settings

> 3. Running force:mdapi:describemetadata to get the describe

Reviewer: ```suggestion
3. Run `force:mdapi:describemetadata` to get the describe
```

> 4. Modifying the registry to include the newly found types

Reviewer: ```suggestion
4. Modify the registry to include the newly found types
```

```shell
yarn update-registry
```

### What the script can't do

> inFolderTypes and types with childXml in their describe are not supported. You **want** to explore the various strategies for those (see the SDR Handbook in this repo) and create NUTs to validate that your types are being handled correctly.

Reviewer: ```suggestion
`inFolderTypes` and types with `childXml` in their describe are not supported. You **want** to explore the various strategies for those (see the [SDR Handbook](../HANDBOOK.md) in this repo) and create NUTs to validate that your types are being handled correctly.
```

:question: Based on the wording, it seems like "describe" is an element. If so, that should probably be back-ticked as well.
:question: It might help for "create NUTs" to link to the *Integration Testing* header?

:warning: Confirm the relative URL for the Handbook.

For those situations, refer to another existing type in the registry that you want yours to behave like.

> If that's confusing, it's a great time to reach out to the CLI team.

Reviewer: Since SDR is used by both CLI and VSCode, I want to confirm that "CLI Team" is the best phrase. 

Also, what's the best avenue to reach out? Maybe GH issue is the right public channel?

## Manual Edits

You can do what the script does yourself. As you work, run `./node_modules/mocha/bin/mocha test/registry/registryValidation.test.ts` to check your entries

> Run `sfdx force:mdapi:describemetadata --json` to get the describe. `>` it to a file or use `|jq` to find your type.

Reviewer: ```suggestion
Run `sfdx force:mdapi:describemetadata --json` to get the describe. `>` the output to a file or pipe it to [jq](https://stedolan.github.io/jq/) (` | jq`) to find your type.
```

Your describe will contains something like this

```json
{
    "directoryName": "yourTypes",
    "inFolder": "false",
    "metaFile": "false",
    "suffix": "yourtype",
    "xmlName": "YourNewType"
},
```

> Entries in the describe.json have the following shape:

Reviewer: ```suggestion
Entries in the `describe.json` have the following shape:
```

| Property | Description|
| --- | --- |
|childXmlNames | List of child sub-components for this component|
|directoryName | The name of the directory in the .zip file that contains this component.|
| inFolder | Indicates whether the component is in a folder (true) or not (false). For example, documents, email templates and reports are stored in folders.|
|metaFile| Indicates whether the component requires an accompanying metadata file. For example, documents, classes, and s-controls are components that require an additional metadata file.|
|suffix| The file suffix for this component.|
|xmlName| The name of the root element in the metadata file for this component. This name also appears in the Packages > types > name field in the manifest file package.xml.|

> You'd create an entry in the `types` property.

Reviewer: ```suggestion
Create an entry in the `types` property:
```

> this fictitious type COULD HAVE used the automated script

```json
"yournewtype": { // the key is lowercased xmlName
      "id": "yournewtype", // lowercased xmlName
      "name": "YourNewType", // xmlName
      "suffix": "yourtype", // suffix
      "directoryName": "yourTypes", //directoryName
      "inFolder": false, // inFolder
      "strictDirectoryName": false // set to true when suffixes are not unique in the registry
    },
```

add a matching entry in `suffixes`

```json
  "yourtype": "yournewtype" // "suffix": "id"
```

### When suffixes aren't unique

> If your type uses a suffix already in use, you'll need to tell SDR to use the directoryName as well by setting `"strictDirectoryName"` : `"true"`

Reviewer: ```suggestion
If your type uses a suffix already in use, tell SDR to use the `directoryName` as well by setting `"strictDirectoryName"` : `"true"`
```

Then also add an entry in the `strictDirectoryNames` section.

```json
  "yourtype" : "yournewtype" // "directory" : "id"

```

To preserve existing behavior, use `strictDirectoryName` on the new types, not the old ones.

# Testing

> SDR include 2 registry-related tests to check your changes to the metdataRegistry.json

Reviewer: ```suggestion
SDR includes 2 registry-related tests to check your changes against the `metdataRegistry.json`
```

:warning: Confirm that the word "against" is appropriate. "Check your changes _to_" was awkward to read.

Test failures here could be types that exist in the `types` section but don't have entries in `suffixes` or `strictDirectoryNames`.
It also checks that suffixes are unique OR only one type that shares a suffix isn't `strictDirectoryName`.

## Validate the registry is complete

The library uses the [registry file](../src/registry/metadataRegistry.json) to resolve how to process metadata types. This needs to be updated as new metadata types are added to the platform at major releases.

The completeness is checked by comparing the registry to the metadata coverage report, but excluding

1. Types that aren't supported in the metadata API
2. Types in the [nonSupportedTypes file](../src/registry/nonSupportedTypes.ts) (think of it as a registry-ignore file). You can ignore the types themselves, or the feature/settings they depend on. Be sure to explain why you're choosing to ignore that type.

## Manual Testing

Want to make sure your types are working as expected?

1. Create a new project with `sfdx force:project:create -n registryTest`
1. Create a scratch org `sfdx force:org:create`
1. Open the org and create your types.
1. Run `sfdx force:source:beta:status` and verify the remote add.

> 1. `sfdx force:source:beta:pull` to pull the metadata and examine what is retrieved

Reviewer: ```suggestion
1. Run `sfdx force:source:beta:pull` to pull the metadata and examine what is retrieved
```

1. Run `sfdx force:source:beta:status` and verify the changes were retrieved and no longer appear.
1. Delete the org `sfdx force:org:delete --noprompt`
1. Create a new scratch org. `sfdx force:org:create`
1. Push the source `sfdx force:source:beta:push`
1. Convert the source to mdapi format `sfdx force:source:convert -d mdapiOut`

> 1. look in the resulting `metadataPackage_` and `package.xml` to see that it looks as expected

Reviewer: ```suggestion
1. Look in the resulting `metadataPackage_` and `package.xml` to see that it looks as expected
```

> 1. deploy it to the org using `sfdx force:mdapi:deploy --deploydir mdapiOut --wait 30` and verify that it succeeds

Reviewer: ```suggestion
1. Deploy it to the org using `sfdx force:mdapi:deploy --deploydir mdapiOut --wait 30` and verify that it succeeds
```

> 1. delete the source directory `rm -rf force-app/main/default/*`

Reviewer: ```suggestion
1. Delete the source directory `rm -rf force-app/main/default/*`
```

1. Create a new scratch org and convert the source back
1. Convert back from mdapi to source format `sfdx force:mdapi:convert -r mdapiOut -d force-app`
1. `sfdx force:source:beta:push`

### Caveats

> Only source:push and source:pull support source tracking so the target types must be MDAPI addressable on the server. If they aren’t then special code is needed to support source tracking for these components. See the document [Metadata API Types: End to End, Cradle to Grave](https://docs.google.com/document/d/13jJLbM_ztS0tOOV8bQzNAnEFbL5MY4rNLa18oDV2SCs/edit?usp=sharing) (Salesforce internal only) for more details.

Reviewer: ```suggestion
Only `source:push` and `source:pull` support source tracking, so the target types must be MDAPI addressable on the server. If they aren’t MDAPI addressable, special code is needed to support source tracking for these components. See the document [Metadata API Types: End to End, Cradle to Grave](https://docs.google.com/document/d/13jJLbM_ztS0tOOV8bQzNAnEFbL5MY4rNLa18oDV2SCs/edit?usp=sharing) (Salesforce internal only) for more details.
```

## Unit Testing

SDR Unit tests are abstract (mockRegistry tries to represent possible patterns). If your type uses a combination of properties that don't match an existing type in the registry, there's a chance the behavior isn't covered.

> Reach out to the CLI team for help with unit tests.

Reviewer: Is there a convenient link for how to reach out? If so, make `CLI Team` a link.

[metadataResolverRealRegistry.ts](../test/resolve/metadataResolverRealRegistry.test.ts) is an example of unit testing one behavior (resolving from source files) of a real metadata type.

## Integration Testing

> If you're doing anything complex (you've used any of the following properties `strategies`, `folderType`, `inFolder=true`, `ignoreParsedFullName`, `folderContentType`, `ignoreParentName`) then you'll wanted to add some NUTs that verify the behavior or your types using real orgs and to prevent SDR changes from causing regressions on your types.

Reviewer: ```suggestion
If you're doing anything complex (you've used any of the following properties `strategies`, `folderType`, `inFolder=true`, `ignoreParsedFullName`, `folderContentType`, `ignoreParentName`), you'll want to add some NUTs that verify the behavior or your types using real orgs and to prevent SDR changes from causing regressions on your types.
```

# Contributing Metadata Types to the Registry

Got questions?

- If you work for Salesforce, [#platform-cli](https://salesforce-internal.slack.com/archives/C01LKDT1P6J).
- If not, [open an issue](https://github.com/forcedotcom/cli/issues)

## Adding new types to the registry via a script

> NOTE: If you have a “Settings” metadata type, you do not need to follow these instructions. Support for your type is already built into SDR.

To simplify modifying the registry, there's a script.

> Note: The script depends on being able to create an org that will have your type. If you have a working org, but your type isn't compatible with scratch org, you can [use an existing org](#use-an-existing-org)

The script will:

1. Look for missing types (similar to the completeness test)
2. For missing types, generate a project and scratch org that includes the Features/Settings
3. Run `force:mdapi:describemetadata` to get the describe
4. Modify the registry to include the newly found types

```shell
yarn update-registry
```

### What the script can't do

> inFolderTypes and types with childXml in their describe are not supported. You **want** to explore the various strategies for those see the [SDR Handbook](../HANDBOOK.md) in this repo) and [create tests](#integration-testing) to validate that your types are being handled correctly.

Reviewer 2: Missing `(` 
`strategies for those see the`
`strategies for those (see the`

|directoryName | The name of the directory in the .zip file that contains this component.|
| inFolder | Indicates whether the component is in a folder (true) or not (false). For example, documents, email templates and reports are stored in folders.|
|metaFile| Indicates whether the component requires an accompanying metadata file. For example, documents, classes, and s-controls are components that require an additional metadata file.|
|suffix| The file suffix for this component.|
|xmlName| The name of the root element in the metadata file for this component. This name also appears in the Packages > types > name field in the manifest file package.xml.|

Create an entry in the `types` property:

> this fictitious type COULD HAVE used the automated script

```json
"yournewtype": { // the key is lowercased xmlName
      "id": "yournewtype", // lowercased xmlName
      "name": "YourNewType", // xmlName
      "suffix": "yourtype", // suffix
      "directoryName": "yourTypes", //directoryName
      "inFolder": false, // inFolder
      "strictDirectoryName": false // set to true when suffixes are not unique in the registry
    },

> ```

Reviewer 2: Indent on the last `}` is off. Also I think using `js` syntax would make those comment "errors" less jarring. 
```js
"yournewtype": { // the key is lowercased xmlName
      "id": "yournewtype", // lowercased xmlName
      "name": "YourNewType", // xmlName
      "suffix": "yourtype", // suffix
      "directoryName": "yourTypes", //directoryName
      "inFolder": false, // inFolder
      "strictDirectoryName": false // set to true when suffixes are not unique in the registry
},
```

1. Look in the resulting `metadataPackage_` and `package.xml` to see that it looks as expected
1. Deploy it to the org using `sfdx force:mdapi:deploy --deploydir mdapiOut --wait 30` and verify that it succeeds
1. Delete the source directory `rm -rf force-app/main/default/*`
1. Create a new scratch org and convert the source back
1. Convert back from mdapi to source format `sfdx force:mdapi:convert -r mdapiOut -d force-app`
1. `sfdx force:source:beta:push`

### Caveats

Only `source:push` and `source:pull` support source tracking, so the target types must be MDAPI addressable on the server. If they aren’t MDAPI addressable, special code is needed to support source tracking for these components. See the document [Metadata API Types: End to End, Cradle to Grave](https://docs.google.com/document/d/13jJLbM_ztS0tOOV8bQzNAnEFbL5MY4rNLa18oDV2SCs/edit?usp=sharing) (Salesforce internal only) for more details.

## Unit Testing

SDR Unit tests are abstract (mockRegistry tries to represent possible patterns). If your type uses a combination of properties that don't match an existing type in the registry, there's a chance the behavior isn't covered.

> Reach out to the CLI team for help with unit tests. .

Reviewer 2: Extra `.`
